packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
-      }
-    };
+        expect(error.message).to.match(/uncle/);
+      });
+
+      const snapshotId = await createSnapshot(simpleStorage.getProvider());
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
 
-    subscription.on("confirmation", onMessage);
+      expect(confirmationFn.called).to.be.true;
 
-    await mineBlocks(simpleStorage.getProvider(), 8);
+      await revertFromSnapshot(simpleStorage.getProvider(), snapshotId);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      // Broadcast same message again (simulate reorg)
+      subscription = simpleStorage.setValue("hello world");
+
+      expect(errorFn.called).to.be.true;
+    });
   });
 
-  it("should throw error when subscribing on invalid event", async () => {
-    const events = ["ValueChanged", "InvalidEvent"];
-    try {
+  describe("contract events subscriptions tests", async () => {
+    beforeEach("assign a wallet to the contract", () => {
+      expect(() => {
+        simpleStorage.setWallet(wallet);
+      }).to.not.throw();
+    });
+
+    it("should throw error when subscribing on invalid event", async () => {
+      const events = ["ValueChanged", "InvalidEvent"];
+
+      expect(() => {
+        subscription = simpleStorage.subscribe(events);
+      }).to.throw();
+    });
+
+    it("should throw error then listening on invalid event", async () => {
+      const events = ["ValueChanged"];
       const subscription = simpleStorage.subscribe(events);
-      assert(false, "subscription with invalid event");
-    } catch (e) {
-      assert(true);
-    }
-  });
 
-  it("should throw error then listening on invalid event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
+      expect(() => {
+        subscription.on("InvalidEvent", () => {});
+      }).to.throw();
+    });
 
-    try {
-      subscription.on("InvalidEvent", () => {});
-      assert(false, "listening with invalid event");
-    } catch (e) {
-      assert(true);
-    }
-  });
+    it("should listen to an event", async () => {
+      const events = ["ValueChanged"];
+      const subscription = simpleStorage.subscribe(events);
+      const fakeFn = sinon.fake();
 
-  it.skip("should listen to an event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
-    const handlerFunction = async message => {};
+      const handlerFunction = message => {
+        const { data } = message;
+        const { args } = data;
+        fakeFn();
+        subscription.removeListener("ValueChanged", handlerFunction);
+      };
+
+      subscription.on("ValueChanged", handlerFunction);
+
+      simpleStorage.setValue("hello world");
+
+      await mineBlocks(simpleStorage.getProvider(), 10);
+
+      expect(fakeFn.called).to.be.true;
+    });
 
-    subscription.on("ValueChanged", handlerFunction);
+    it("should remove listener after timeout", async () => {
```

Comment:
I think listening for an event should be eternal by default. If they want it go away, in general the user should remove it manually. We could add a convenience function for subscribing with a pre-set timeout either for that event subscription or all event subscriptions, though. But I don't think that's a high priority.

Handling no confirmations of a transaction after a certain amount of time is a different story, though.

Reply:
A. Automaticly listener removal was resolved on PR https://github.com/tasitlabs/TasitSDK/pull/102.
B. TransactionSubscription has now a `setEventsTimeout` method.
C. About time-out policy, I've made a question on thread https://github.com/tasitlabs/TasitSDK/pull/87#discussion_r245880280

Reply:
That all sounds good. Since I responded in that other thread on topic C, I'll resolve this.